### PR TITLE
BUG: Fix return shape of inverse_indices in unique_inverse

### LINKED ISSUE #25552: BUG: `np.unique_inverse` output has wrong shape
### Describe the issue:

According to the [`unique_inverse` specification](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.unique_inverse.html), the `inverse_indices` output of `np.unique_inverse` should have the same shape as the input:

> second element must have the field name `inverse_indices` and must be an array containing the indices of values that reconstruct `x`. The array must have the same shape as `x` and have the default array index data type.

In NumPy's current implementation, `inverse_indices` is a flattened array.

A similar issue exists for `numpy.unique_all`.

### Reproduce the code example:

```python
import numpy as np
x = np.ones((2, 3))
out = np.unique_inverse(x)
assert out.inverse_indices.shape == x.shape, f"{out.inverse_indices.shape=} {x.shape=}"
```

### Error message:

```shell
AssertionError: out.inverse_indices.shape=(6,) x.shape=(2, 3)
```

### Python and NumPy Versions:
```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.0.0.dev0+git20240106.42fef42
3.9.6 (default, May  7 2023, 23:32:44) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]
```
### Runtime Environment:
```
>>> import numpy; print(numpy.show_runtime())
[{'numpy_version': '2.0.0.dev0+git20240106.42fef42',
  'python': '3.9.6 (default, May  7 2023, 23:32:44) \n'
            '[Clang 14.0.3 (clang-1403.0.22.14.1)]',
  'uname': uname_result(system='Darwin', node='jmdg-macbookpro.roam.internal', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #25552

### CODE REVIEW

**Comment by seberg** [comment]
Should we also just change this in the normal version?  It does look much like a bug-fix when `axis` is used?

**Comment by jakevdp** [comment]
I don't think we should make this change in `np.unique`, because that's a potentially breaking change.

`unique_all` and `unique_inverse` have not been part of a release yet, so modifying their output shapes is safer.

**Comment by seberg** [comment]
No strong opinion, although it seems so odd that I think we can do it, and no time like a 2.0 release to change such a thing that looks like an oversight to begin with.  (If users worked around via `.reshape(input.shape)` their code even keeps working and `res.reshape(-1)` gives the old without branching)

**Comment by jakevdp** [comment]
Sure, I see what you're saying. I'm happy to update the PR to modify the original (and add appropriate docs/changelog) if you'd prefer that.

**Review by mhvk — COMMENTED** [review]
Good to fix this! 

Like @seberg, I think we should ideally fix this in `np.unique` itself, since it is clearly a bug.

That said, `np.unique` has an `axis` argument, and with that the simple reshape is no longer generally correct. It would need something like the following:
```
arr = np.array([[[1, 2, 3], [2, 3, 1]]]*2 +[[[1, 3, 2], [2, 1, 3]]])
axis = 1  # for testing
a, i = np.unique(arr, axis=axis, return_inverse=True)
if axis is None:
    i.shape = arr.shape
else:
    i.shape = tuple((sh if ax == axis else 1) for ax, sh in enumerate(arr.shape))
    # we probably want this too, to make the indices similar to argsort.
    i = np.broadcast_to(i, arr.shape)

# check result is correct, in way also suggested for np.argsort, etc.
np.all(np.take_along_axis(a, i, axis=axis) == arr)
# all true
```

File: numpy/lib/tests/test_arraysetops.py [hunk 1]
```diff
@@ -950,3 +950,13 @@ def test_unique_array_api_functions(self):
...
+
+    def test_unique_inverse_shape(self):
+        # Regression test for https://github.com/numpy/numpy/issues/25552
+        arr = np.array([[1, 2, 3], [2, 3, 1]])
+        expected_values, expected_inverse = np.unique(arr, return_inverse=True)
+        expected_inverse = expected_inverse.reshape(arr.shape)
+        for func in np.unique_inverse, np.unique_all:
+            result = func(arr)
+            assert_array_equal(expected_values, result.values)
+            assert_array_equal(expected_inverse, result.inverse_indices)
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply numpy/lib/tests/test_arraysetops.py:962]: Done!

**Comment by jakevdp** [comment]
I think this change to `np.unique` makes sense, but maybe it deserves its own discussion? Should we merge this PR (so that `unique_*` have the correct API) and then discuss `np.unique` in its own issue and/or mailing list thread?

**Comment by mattip** [comment]
In the triage meeting we reached a consensus to call the wrong 1d shape a bug, and fix it without any deprecation. It should be mentioned in the release notes.

**Comment by jakevdp** [comment]
Thanks - I will update this PR to change the output shape for `np.unique` as well.

File: numpy/lib/_arraysetops_impl.py [hunk 2]
```diff
@@ -222,6 +222,12 @@ def unique(ar, return_index=False, return_inverse=False,
         lexicographical order is chosen - see np.sort for how the lexicographical
         order is defined for complex arrays.
 
+    .. versionchanged: NumPy 2.0
```
Comment (mhvk) [inline numpy/lib/_arraysetops_impl.py:225]: Remove "numpy" here.

**Review by mhvk — APPROVED** [review]
Nice! Only a silly nitpick from me. Approving.

**Comment by mhvk** [comment]
Note there is a linting error, so you can fix the versionchanged at the same time...

File: numpy/lib/_arraysetops_impl.py (hunk 2 above)
Comment (jakevdp) [reply numpy/lib/_arraysetops_impl.py]: Done, also removed it a few lines above from the line I copied this from 😁

Comment (mhvk) [reply numpy/lib/_arraysetops_impl.py]: Ah, yes, that explains it... thanks for fixing that.

**Review by mhvk — APPROVED** [review]
Thanks again!

**Comment by rgommers** [comment]
This broke SciPy in two places, in ways that were a little nontrivial to diagnose: https://github.com/scipy/scipy/pull/19868. The fix does make sense, but more complaints may roll in - this probably broke most usages of the returned reverse indices.

**Comment by mhvk** [comment]
Hmm, @jakevdp was clearly right to worry about breaking things. Arguably in cases like this we should run the scipy tests... Thanks for fixing!

**Comment by jakevdp** [comment]
@rgommers do you think this change should be reverted? I do think the change in the `axis=None` case is a good one, but I don't feel as strongly about the change of behavior when `axis` is an int.

**Comment by rgommers** [comment]
I don't know, I'd give it a week or so - if no issues turn up for other libraries that test against numpy nightlies, it's probably okay to keep the change.

**Comment by WarrenWeckesser** [comment]
One more data point: this broke some code in an unreleased project of mine ([`yanova`](https://github.com/WarrenWeckesser/yanova)).  Some tests fail because of the change of shape of the inverse array in examples like the following.

NumPy 1.26.4:

```
In [9]: np.__version__
Out[9]: '1.26.4'

In [10]: m = np.array([[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2],
    ...:               [0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0]])

In [11]: np.unique(m, axis=1, return_inverse=True)
Out[11]: 
(array([[0, 0, 1, 1, 2, 2],
        [0, 1, 0, 1, 0, 1]]),
 array([0, 0, 0, 1, 1, 1, 3, 3, 2, 2, 5, 5, 5, 4, 4]))
```

NumPy 2.0.0rc2

```
In [6]: np.__version__
Out[6]: '2.0.0rc2'

In [7]: m = np.array([[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2],
   ...:               [0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0]])

In [8]: np.unique(m, axis=1, return_inverse=True)
Out[8]: 
(array([[0, 0, 1, 1, 2, 2],
        [0, 1, 0, 1, 0, 1]]),
 array([[0, 0, 0, 1, 1, 1, 3, 3, 2, 2, 5, 5, 5, 4, 4]]))
```

I can easily fix the issue, and since we haven't heard of any other problems related to this, this doesn't open up again the question of reverting the change.  Just FYI.

**Comment by seberg** [comment]
Well, I thought the reason why this was different in the Array API was because the current behavior doesn't make sense...
... so that is a bit my bad, I didn't think much about it assuming that the Array API decision was base on the current version being unwieldy.

If that was the case, I would not want to revert this.  But, unfortunately, that impression was seems wrong.  There is no big reason for `unique` to behave differently, because `unique` is already compatible to `np.take` rather than `np.take_along_axes`, and that isn't going away.

**Comment by seberg** [comment]
Just FYI, this is now reverted except for `axis=None` where the change made sense.  For `axis=0`, etc. I think the old behavior was actually strictly better anyway.

**Pull request merged**